modules/configuration/src/test/java/org/apache/ignite/internal/configuration/util/ConfigurationUtilTest.java

```diff
@@ -808,6 +809,26 @@ void testRemoveLastKey() {
         assertEquals(List.of("0"), removeLastKey(List.of("0", "1")));
     }
 
+    /**
+     * Tests that {@link ConfigurationUtil#addDefaults} copies the tree when adding default values.
+     */
+    @Test
+    void testAddDefaultsPurity() {
+        InnerNode parentNode = newNodeInstance(ParentConfigurationSchema.class);
+
+        ParentChange parentChange = (ParentChange) parentNode;
+
+        parentChange.changeElements(elements -> elements.create("name", element -> {}));
+
+        NamedListNode<?> beforeDefaults = find(List.of("elements"), parentNode, true);
```

Comment:
`find(List.of("elements"), parentNode, true)`
This is equivalent to `parentChange.elements()`